compiler-cli/src/lsp_fs_proxy.rs

```diff
@@ -0,0 +1,218 @@
+        self.project_io.hardlink(from, to)
+    }
+
+    fn symlink_dir(&self, from: &Path, to: &Path) -> Result<(), Error> {
+        self.project_io.symlink_dir(from, to)
+    }
+
+    fn delete_file(&self, path: &Path) -> Result<(), Error> {
+        self.project_io.delete_file(path)
+    }
+}
+
+impl FileSystemReader for LspFsProxy {
+    // Read from disk since `LspFsProxy` should never produce any new files
+    fn gleam_source_files(&self, dir: &Path) -> Vec<PathBuf> {
+        self.project_io.gleam_source_files(dir)
+    }
+
+    // Read from disk since `LspFsProxy` should never produce any new files
+    fn gleam_cache_files(&self, dir: &Path) -> Vec<PathBuf> {
+        self.project_io.gleam_cache_files(dir)
+    }
+
+    // Read from disk since `LspFsProxy` never creates any dirs
+    fn read_dir(&self, path: &Path) -> Result<ReadDir> {
+        self.project_io.read_dir(path)
+    }
+
+    fn read(&self, path: &Path) -> Result<String, Error> {
+        // Note that files are assumed to be stored under abs-path keys
+        let in_mem_result = self.cache.read(abs_path(path)?.as_path());
+        match in_mem_result {
+            Ok(_) => {
+                tracing::info!("Reading file from cache: {}", path.to_string_lossy());
+                in_mem_result
+            }
+            Err(e) => {
+                tracing::info!(
+                    "Got {} => Reading file from disk: {}",
+                    e,
+                    path.to_string_lossy()
+                );
+                self.project_io.read(path)
+            }
+        }
+    }
+
+    fn read_bytes(&self, path: &Path) -> Result<Vec<u8>, Error> {
+        // Note that files are assumed to be stored under abs-path keys
+        let in_mem_result = self.cache.read_bytes(abs_path(path)?.as_path());
+        match in_mem_result {
+            Ok(_) => {
+                tracing::info!(
+                    "Reading (bytes) file from cache: {}",
+                    path.to_string_lossy()
+                );
+                in_mem_result
+            }
+            Err(e) => {
+                tracing::info!(
+                    "Got {} => Reading file from disk: {}",
+                    e,
+                    path.to_string_lossy()
+                );
+                self.project_io.read_bytes(path)
+            }
+        }
+    }
+
+    // Not implemented in `InMemoryFileSystem`
+    fn reader(&self, path: &Path) -> Result<WrappedReader, Error> {
+        self.project_io.reader(path)
+    }
+
+    // Read from disk, all files on disk are also files in mem-cache
+    fn is_file(&self, path: &Path) -> bool {
+        self.project_io.is_file(path)
+    }
+
+    // Read from disk - mem cache has no dirs
+    fn is_directory(&self, path: &Path) -> bool {
+        self.project_io.is_directory(path)
+    }
+
+    // Not applicable for mem-cache
+    fn current_dir(&self) -> Result<PathBuf, Error> {
+        self.project_io.current_dir()
+    }
```

Comment:
Could you call the in-memory implementation for these please 🙏

Reply:
Sure! I left them out because they were not really relevant in relation to how the `InMemoryFileSystem` is used by `LspFsProxy`, but I can add them if you think they are needed. Maybe `reader` will be a bit more work since it is not implemented by `InMemoryFileSystem` today.

Reply:
Let's just have that one crash for now :)

Reply:
:)
Alright, I can do that (assuming that there is actually no call to it during LSP compilation). Otherwise maybe it could be possible to just implement a reader that buffers the content of a normal read (from the cache).

Reply:
Actually `current_dir` will also be a problem I think. The current implementation in mem-cache is to just return "/" - so its hard to do a fallback logic.

`is_file` / `is_directory` can be done though assuming that we fallback to disk only if the mem-cache returns false. Meaning if the file/dir exists in mem-cache it will be considered to be existing even if it was removed from disk.

Reply:
OK! Thanks for digging into it. Let's skip the current dir one then for now.

Reply:
I think we should leave it for now. An implementation we would include wouldn't panic for multiple references (which there will be) or use indexing.

Reply:
Yes, that makes sense - I'll remove it.